core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtils.scala

```diff
@@ -356,4 +374,233 @@ private[mesos] trait MesosSchedulerUtils extends Logging {
     sc.conf.getTimeAsSeconds("spark.mesos.rejectOfferDurationForReachedMaxCores", "120s")
   }
 
+  /**
+   * Checks executor ports if they are within some range of the offered list of ports ranges,
+   *
+   * @param sc the Spark Context
+   * @param ports the list of ports to check
+   * @return true if ports are within range false otherwise
+   */
+  protected def checkPorts(sc: SparkContext, ports: List[(Long, Long)]): Boolean = {
+
+    def checkIfInRange(port: Long, ps: List[(Long, Long)]): Boolean = {
+      ps.exists(r => r._1 <= port & r._2 >= port)
+    }
+
+    val portsToCheck = ManagedPorts.getPortValues(sc.conf)
+    val nonZeroPorts = portsToCheck.filter(_ != 0)
+    val withinRange = nonZeroPorts.forall(p => checkIfInRange(p, ports))
+    // make sure we have enough ports to allocate per offer
+    ports.map(r => r._2 - r._1 + 1).sum >= portsToCheck.size && withinRange
+  }
+
+  /**
+   * Partitions port resources.
+   *
+   * @param conf the spark config
+   * @param ports the ports offered
+   * @return resources left, port resources to be used and the list of assigned ports
+   */
+  def partitionPorts(
+      conf: SparkConf,
+      ports: List[Resource])
+    : (List[Resource], List[Resource], List[Long]) = {
+    val taskPortRanges = getRangeResourceWithRoleInfo(ports.asJava, "ports")
+    val portsToCheck = ManagedPorts.getPortValues(conf)
+    val nonZeroPorts = portsToCheck.filter(_ != 0)
+    // reserve non zero ports first
+    val nonZeroResources = reservePorts(taskPortRanges, nonZeroPorts)
+    // reserve actual port numbers for zero ports - not set by the user
+    val numOfZeroPorts = portsToCheck.count(_ == 0)
+    val randPorts = pickRandomPortsFromRanges(nonZeroResources._1, numOfZeroPorts)
+    val zeroResources = reservePorts(nonZeroResources._1, randPorts)
+    val (portResourcesLeft, portResourcesToBeUsed) =
+      createResources(nonZeroResources, zeroResources)
+    (portResourcesLeft, portResourcesToBeUsed, nonZeroPorts ++ randPorts)
+  }
+
+  private object ManagedPorts {
```

Comment:
It just groups logically some data and an operation on them. Otherwise i have to put the data (portNames ) as a top level variable and getPortValues  as a top level method. With the latter everything is put inside a top level trait without any straightforward correlation.